notifier/notifier.go

```diff
@@ -266,7 +266,10 @@ func (n *Manager) ApplyConfig(conf *config.Config) error {
 		if err != nil {
 			return err
 		}
-
+		// If old ams exist, use old ams to avoid empty ams when reload.
+		if oldAms, ok := n.alertmanagers[k]; ok {
```

Comment:
Unfortunately ToMap does not guaratee that the configuration of the alertmanager is the same. The key is just  `fmt.Sprintf("config-%d", i)] = a[i]`, e.g. config-1.                                
 so if we change the AM config completely, we would end up with a wrong AM.

Reply:
@roidelapluie  Oh,I see.

If we diff alertmanagerConfigs here, that's ok ? like this:

```go
// ApplyConfig updates the status state as the new config requires.
func (n *Manager) ApplyConfig(conf *config.Config) error {
	n.mtx.Lock()
	defer n.mtx.Unlock()

	n.opts.ExternalLabels = conf.GlobalConfig.ExternalLabels
	n.opts.RelabelConfigs = conf.AlertingConfig.AlertRelabelConfigs
       
	// Diff alert config
       ...
        if oldAlertConf == newAlertConf {
            return nil
        }
        ...
}
```

Reply:
I do not think this could work, we would need to be somehow more clever.

I am wondering if that is causing problems in practice, because it might be done on purpose in the Prometheus server. At least in the past this might help reloading CA files to reach service discoveries. I wonder if there are edge cases like this that would require to reload every time to be correct.